src/storage/src/sink/healthcheck.rs

```diff
@@ -0,0 +1,715 @@
+                            &actual_upper
+                        );
+                        // Sync to the new upper, go to the loop again
+                        self.sync(&actual_upper.0).await;
+                        // If we can't transition to the new status after the sync, no need to do anything else
+                        if !self.current_status.can_transition(&status_update) {
+                            break;
+                        }
+                    }
+                    Ok(Err(invalid_use)) => panic!("compare_and_append failed: {invalid_use}"),
+                    // An external error means that the operation might have succeeded or failed but we
+                    // don't know. In either case it is safe to retry because:
+                    // * If it succeeded, then on retry we'll get an `Upper(_)` error as if some other
+                    //   process raced us. This is safe and will just cause the healthchecker to sync
+                    //   again, and on retry it will notice that the new state was already processed and
+                    //   finish successfully.
+                    // * If it failed, then we'll succeed on retry and proceed normally.
+                    Err(external_err) => {
+                        trace!("compare_and_append in update_status failed: {external_err}");
+                        continue;
+                    }
+                };
+            }
+        }
+    }
+
+    /// Synchronizes internal state with state in the storage collection up until a given timestamp
+    async fn sync(&mut self, target_upper: &Antichain<Timestamp>) {
+        while PartialOrder::less_than(&self.upper, target_upper) {
+            for event in self.listener.next().await {
+                match event {
+                    ListenEvent::Progress(new_upper) => {
+                        self.upper = new_upper;
+                    }
+                    ListenEvent::Updates(updates) => {
+                        self.process_collection_updates(updates);
+                    }
+                }
+            }
+        }
+    }
+
+    /// Bootstraps the state of this Healthchecker instance by reading data from the
+    /// underlying storage collection
+    ///
+    /// This function works by first reading a snapshot of the collection at its `since`,
+    /// and then using the listener to read all updates from `since` up until (but not including)
+    /// `upper`. This is done as a way to read all data in the collection, but without
+    /// having to assume that the `upper` is a single `u64`.
+    async fn bootstrap_state(
+        &mut self,
+        mut read_handle: ReadHandle<SourceData, (), Timestamp, i64>,
+        upper: &Antichain<Timestamp>,
+    ) {
+        let since = read_handle.since().clone();
+        trace!("Bootstrapping state as of {:?}!", since);
+        // Ensure the collection is readable at `since`
+        if PartialOrder::less_than(&since, &self.upper) {
```

Comment:
This still has the bug that I proposed a fix to in that other branch, before I decided I wasn't going to use this implementation in the source. (`self.upper` is always `minimum` here, so this branch is never taken, and we always ignore the content of the state.)

I'm okay merging this because this code is never invoked in this branch, but we should make sure to replace this implementation by the time we enable things.